### ISSUE TITLE
Update to TS 4.4 and update types to support exactOptionalPropertyTypes

### ISSUE THREAD

**Issue by merrywhether** [relay-tools/relay-compiler-language-typescript#447]
[TypeScript 4.4 introduces](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-4.html#exact-optional-property-types---exactoptionalpropertytypes) the `exactOptionalPropertyTypes` [compiler option](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes). This change means that optional properties are no longer conflated with properties whose value could be `undefined`.

This comes up in regards to optional query parameters which are currently typed as
```ts
foo?: Foo | null; // which is treated as `Foo | null | undefined` without this flag
```

It's not uncommon for such parameters to receive variables who themselves are also `| undefined`, especially via optional propagation (this is a strained example for simplicity):
```ts
function fakeMakeQuery(foo?: QueryVariables["foo"]) {
  const variables: QueryVariables = {
    foo: foo, // with this flag enabled, TS complains about "undefined can't be assigned to Foo | null"
  };
  ...
}
```
You can workaround this by adding extra logic like nullish coalescing (`foo: foo ?? null`) or if-gated property addition
```ts
const variables: QueryVariables = {};
if (foo) {
  variables.foo = foo;
}
```
But these both incur runtime costs.

Desired change:
Optional parameters should expand their type union to allow consumers to use the `exactOptionalPropertyTypes` flag by changing from
```ts
foo?: Foo | null;
```
to
```ts
foo?: Foo | null | undefined;
```
